### PR TITLE
Only one step object

### PR DESCRIPTION
**Pull request by dirkgr**
#544 needs to be merged first.

We call `deepcopy()` on params objects, which contain Step objects, to set `raw_config`. That makes a copy of the step object every time it is used. Worse, this happens recursively, so we make a copy of a step object for every downstream step that depends on it directly or indirectly.

This solution replaces `Step` with refs when we're constructing the raw configs.

### CODE REVIEW

**Comment by dirkgr** [comment]
It would not be pretty. We'd have to go through the arguments to the step and copy them manually, in the style of `_replace_step_dependencies()`. But it's possible.

**Comment by epwalsh** [comment]
> It would not be pretty. We'd have to go through the arguments to the step and copy them manually, in the style of `_replace_step_dependencies()`. But it's possible.

Hmm, alright, I'll leave it up to you then. If you think this is the best way, that's fine with me.

File: tango/step.py [hunk 1]
```diff
@@ -432,6 +432,14 @@ def from_params(  # type: ignore[override]
 
         return subclass(step_name=step_name, step_config=raw_step_config, **kwargs)
 
+    # We only want one step object per step with a given unique id.
+    def __copy__(self):
+        return self
+
+    # We only want one step object per step with a given unique id.
+    def __deepcopy__(self, memo=None):
```
Comment (AkshitaB) [inline tango/step.py:440]: Can we put a log warning that this is being overridden? I don't like that the expected behavior is being changed quietly.

**Review by AkshitaB — COMMENTED** [review]
Will this have unexpected behavior when step objects themselves contain some dictionaries, etc? Maybe won't matter as much when config files are used, but if someone is defining something programmatically?

**Review by AkshitaB — APPROVED** [review]

**Pull request merged**